drivers/pinctrl/intel/pinctrl-intel.c

```diff
@@ -976,6 +991,49 @@ static int intel_gpio_irq_wake(struct irq_data *d, unsigned int on)
 	return 0;
 }
 
+/* Interrupt would sometimes stop triggering for unknown reason. Add a timer
+ * to re-enable the interrupt, then the registered ISR would have chance to
+ * process the pending interrupt.
+ */
+static void intel_gpio_community_int_checker(unsigned long arg)
+{
+	struct intel_pinctrl *pctrl = (struct intel_pinctrl *)arg;
+	int i, gpp;
+
+	for (i = 0; i < pctrl->ncommunities; i++) {
+		struct intel_community *community;
+		community = &pctrl->communities[i];
+
+		if (!(community->features & PINCTRL_FEATURE_INT_CHECK_TIMER))
+			continue;
+
+		for (gpp = 0; gpp < community->ngpps; gpp++) {
+			const struct intel_padgroup *padgrp = &community->gpps[gpp];
+			unsigned long pending, enabled;
+			void __iomem *base = community->regs;
+
+			pending = readl(base + GPI_IS + padgrp->reg_num * 4);
+			enabled = readl(base + community->ie_offset +
+					padgrp->reg_num * 4);
+
+			/* Only interrupts that are enabled */
+			pending &= enabled;
+
+			if (pending) {
+				unsigned long flags;
+
+				dev_dbg(pctrl->dev, "re-enable interrupt for pin %d in timer\n", padgrp->reg_num);
+				raw_spin_lock_irqsave(&pctrl->lock, flags);
+				writel(~pending, base + community->ie_offset + padgrp->reg_num * 4);
+				writel(pending, base + community->ie_offset + padgrp->reg_num * 4);
+				raw_spin_unlock_irqrestore(&pctrl->lock, flags);
+			}
+		}
+	}
+
+	mod_timer(&pctrl->stall_int_check_timer, jiffies + msecs_to_jiffies(100));
```

Comment:
Can you make the real IRQ handler also do this mod_timer call to postpone the execution by 100ms from that point?

The idea being that we will only execute this timer function if 100ms passes without us receiving a real hardware interrupt. If the hardware interrupt is firing OK, then there is no need to execute the timer.